### PR TITLE
[DEV-2909] New KYC steps

### PR DESCRIPTION
**Pull request by lapatric**
### Changes 

- [DEV-2909](https://dfxswiss.atlassian.net/browse/DEV-2909) - new KYC steps 

#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

[DEV-2909]: https://dfxswiss.atlassian.net/browse/DEV-2909?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 1]
```diff
@@ -511,12 +535,24 @@ export class KycService {
...
         return { nextStep: { name: nextStep, preventDirectEvaluation }, nextLevel: KycLevel.LEVEL_10 };
 
+      case KycStepName.LEGAL_ENTITY:
+        return { nextStep: { name: nextStep, preventDirectEvaluation } };
+
+      case KycStepName.STOCK_REGISTER:
+        return { nextStep: { name: nextStep, preventDirectEvaluation } };
+
       case KycStepName.NATIONALITY_DATA:
         return { nextStep: { name: nextStep, preventDirectEvaluation }, nextLevel: KycLevel.LEVEL_20 };
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:545]: Looks like we need to improve the `nextLevel` thing 🤔  Level 20 should be set when completing personal data.

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.entity.ts:1]: Add migration

File: src/subdomains/generic/kyc/enums/kyc.enum.ts [hunk 2]
```diff
@@ -17,24 +26,21 @@ export function getKycStepIndex(stepName: KycStepName): number {
...
-        KycStepName.FINANCIAL_DATA,
-        KycStepName.DFX_APPROVAL,
-      ];
+  return [
+    KycStepName.CONTACT_DATA,
+    KycStepName.PERSONAL_DATA,
+    userData.accountType === AccountType.BUSINESS ? KycStepName.LEGAL_ENTITY : null,
+    userData.legalEntity === LegalEntity.PUBLIC_LIMITED_COMPANY ? KycStepName.STOCK_REGISTER : null,
+    KycStepName.NATIONALITY_DATA,
+    [AccountType.BUSINESS, AccountType.BUSINESS].includes(userData.accountType)
```
Comment (Yannick1712) [inline src/subdomains/generic/kyc/enums/kyc.enum.ts:35]: Do you mean `AccountType.SOLE_PROPRIETORSHIP`?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 3]
```diff
@@ -309,6 +320,19 @@ export class KycService {
...
+
+    user = await this.userDataService.updateUserDataInternal(user, data);
+
+    user = user.internalReviewStep(kycStep);
+    await this.createStepLog(user, kycStep);
+    await this.updateProgress(user, false);
+
+    return KycStepMapper.toKycResult(kycStep);
+  }
+
```
Comment (Yannick1712) [inline src/subdomains/generic/kyc/services/kyc.service.ts:335]: Merge `updateSignatoryPowerData`, `updateNationalityData` and `updateLegalEntityData` together in one method. It's the same logic for all 3 methods, or?

Comment (lapatric) [reply src/subdomains/generic/kyc/services/kyc.service.ts:335]: I agree, makes sense 👍

File: src/subdomains/generic/kyc/enums/kyc.enum.ts (hunk 2 above)
Comment (Yannick1712) [reply src/subdomains/generic/kyc/enums/kyc.enum.ts:35]: Ok but I mean the `[AccountType.BUSINESS, AccountType.BUSINESS]`, so two time the same value in the array makes no sense I think, or? 😅

File: src/subdomains/generic/kyc/services/kyc.service.ts (hunk 3 above)
Comment (lapatric) [reply src/subdomains/generic/kyc/services/kyc.service.ts:335]: Would it make sense to do the same inside the `kyc.controller` or is it cleaner to keep things separate there?

Comment (Yannick1712) [reply src/subdomains/generic/kyc/services/kyc.service.ts:335]: Yeah thought about that too. I'm not sure if it makes more sense like that in kyc controller for `outside` devs to understand 🤔 maybe @davidleomay has an opinion on that. Therefore in the controller I would be fine with both ways ✌️

Comment (davidleomay) [reply src/subdomains/generic/kyc/services/kyc.service.ts]: I would keep it in the controller 🙂

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 4]
```diff
@@ -271,33 +266,34 @@ export class KycService {
...
+    requiresInternalReview: boolean,
+  ): Promise<KycResultDto> {
     let user = await this.getUser(kycHash);
     const kycStep = user.getPendingStepOrThrow(stepId);
 
-    const updatedUser = await this.userDataService.updateUserDataInternal(user, data);
+    user = await this.userDataService.updateUserDataInternal(user, data);
 
-    user = updatedUser.internalReviewStep(kycStep);
+    user = requiresInternalReview ? user.internalReviewStep(kycStep) : user.completeStep(kycStep, data);
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:280]: I think we can use the data also for internalReview `internalReviewStep(kycStep, data)`

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 5]
```diff
@@ -506,17 +502,29 @@ export class KycService {
 
     switch (nextStep) {
       case KycStepName.CONTACT_DATA:
-        return { nextStep: { name: nextStep, preventDirectEvaluation } };
+        return { nextStep: { name: nextStep, preventDirectEvaluation }, nextLevel: KycLevel.LEVEL_10 };
 
       case KycStepName.PERSONAL_DATA:
-        return { nextStep: { name: nextStep, preventDirectEvaluation }, nextLevel: KycLevel.LEVEL_10 };
+        return { nextStep: { name: nextStep, preventDirectEvaluation }, nextLevel: KycLevel.LEVEL_20 };
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:508]: Fix logic

**Review by davidleomay — APPROVED** [review]

**Pull request merged**